robot_skills/src/robot_skills/perception.py

```diff
@@ -184,11 +189,83 @@ def detect_faces(self, image=None, stamp=False):
         """
         if not image:
             image = self.get_image()
+        # ToDo: create a consistent API: either return a tuple, a message or a list
         if stamp:
             return self._get_faces(image).recognitions, image.header.stamp
         else:
             return self._get_faces(image).recognitions
 
+    def detect_operator_face(
+        self,
+        image: typing.Optional[Image] = None,
+        expected_operator_position: typing.Optional[kdl.Vector] = None,
+        operator_distance_threshold: typing.Optional[float] = 0.5,
+    ) -> Recognition:
+        """
+        Snap an image with the camera and return the detected face closest to the expected position.
+
+        N.B.: this seems more reliable than using the size of the ROI, which might be affected by, e.g., a big hairdo.
+
+        :param image: image to use for recognition
+        :param expected_operator_position: expected operator position w.r.t. robot base
+        :param operator_distance_threshold: people outside this radius from the expected position are discarded
+        :return: image_recognition_msgs/Recognition
+        :raises: RuntimeError
+        """
+        recognitions = self.detect_faces(image)
+        if not recognitions:
+            raise RuntimeError("No faces detected")
+        operator_recognition = self._filter_operator_recognition(
+            recognitions, expected_operator_position, operator_distance_threshold
+        )
+        return operator_recognition
+
+    def _filter_operator_recognition(
+        self,
+        recognitions: typing.List[Recognition],
+        expected_operator_pos: typing.Optional[kdl.Vector] = None,
+        threshold: float = 0.5,
+    ) -> Recognition:
+        """
+        For all provided recognitions, project the ROI and determine the one closest to the expected operator position,
+        computed on the floor plane.
+
+        :param recognitions: recognitions
+        :param expected_operator_pos: expected position of the operator w.r.t. the robot
+        :param threshold: if the distance between the recognition and the expected pos exceed the threshold, an error
+        is raised
+        :return: Recognition closed to the expected position
+        :raises: RuntimeError
+        """
+        frame_id = f"{self.robot_name}/base_link"
+        projected_recognitions = [(rcg, self.project_roi(rcg.roi, frame_id)) for rcg in recognitions]
+        expected_operator_pos = kdl.Vector(1.0, 0.0, 0.0) if expected_operator_pos is None else expected_operator_pos
```

Comment:
You could use the 1,0,0 vector as default instead of None in that case. That would also make it clear in the functions signature what the default behaviour is.

Reply:
A `kdl.Vector` is mutable. Hence, though that suggestion would work in this specific case, I don' t consider it good programming practice. Consider the following example:
```
In [1]: import PyKDL as kdl                                                                          

In [2]: def foo(bar = kdl.Vector()): 
   ...:     print(f"\nBefore: {bar}") 
   ...:     bar.x(bar.x() + 1.0) 
   ...:     print(f"After: {bar}") 
   ...:                                                                                              

In [3]: foo()                                                                                        

Before: [           0,           0,           0]
After: [           1,           0,           0]

In [4]: foo()                                                                                        

Before: [           1,           0,           0]
After: [           2,           0,           0]
```